subcommands/targets/offline-update.go

```diff
@@ -0,0 +1,286 @@
+	targetName := args[0]
+	dstDir := args[1]
+
+	if len(tag) == 0 {
+		subcommands.DieNotNil(errors.New("missing mandatory flag `--tag`"))
+	}
+
+	fmt.Printf("Checking whether Target exists; target: %s, tag: %s, production: %v\n", targetName, tag, prod)
+	ok, err := checkIfTargetExists(targetName)
+	subcommands.DieNotNil(err, "Failed to check whether Target exists:")
+	if !ok {
+		subcommands.DieNotNil(fmt.Errorf("The specified Target has not been found; target: %s, tag: %s, production: %v\n", targetName, tag, prod))
+	}
+
+	params, err := getTargetInfo(targetName)
+	subcommands.DieNotNil(err, "Failed to obtain Target's details:")
+
+	fmt.Printf("Downloading offline update content of Target %s to %s\n", targetName, dstDir)
+
+	fmt.Println("Downloading TUF metadata...")
+	err = downloadTufRepo(factory, path.Join(dstDir, "tuf"))
+	subcommands.DieNotNil(err, "Failed to download TUF metadata:")
+
+	fmt.Printf("Downloading an ostree repo from the Target's OE build %d...\n", params.ostreeVersion)
+	err = downloadOstree(params.ostreeVersion, params.hardwareID, dstDir)
+	subcommands.DieNotNil(err, "Failed to download Target's ostree repo:")
+
+	fmt.Printf("Downloading Apps fetched by the `assemble-system-image` run; build number:  %d, tag: %s...\n",
+		params.version, params.buildTag)
+	err = downloadApps(targetName, params.version, params.buildTag, path.Join(dstDir, "apps"))
+	subcommands.DieNotNil(err, "Failed to download Target's Apps:")
+
+	fmt.Println("Successfully downloaded offline update content")
+}
+
+func checkIfTargetExists(targetName string) (bool, error) {
+	data, err := api.TufMetadataGet(factory, "targets.json", tag, prod)
+	if herr := client.AsHttpError(err); herr != nil && herr.Response.StatusCode == 404 {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	targets := client.AtsTufTargets{}
+	err = json.Unmarshal(*data, &targets)
+	if err != nil {
+		return false, err
+	}
+	for tn := range targets.Signed.Targets {
+		if tn == targetName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func getTargetInfo(targetName string) (*targetInfo, error) {
+	// Getting Target's custom info from the `/ota/factories/<factory>/targets/<target-name>` because:
+	// 1. a target name is unique and represents the same Target across all "tagged" targets set including prod;
+	// 2. only this target version/representation contains an original tag(s)/branch that
+	// the `image-assemble` and apps fetching was performed for (needed for determining where to download Apps from).
+	custom, err := getTargetCustomInfo(targetName)
+	if err != nil {
+		return nil, err
+	}
+
+	info := targetInfo{}
+	info.version, err = strconv.Atoi(custom.Version)
+	if err != nil {
+		return nil, err
+	}
+	info.hardwareID = custom.HardwareIds[0]
+	info.buildTag = custom.Tags[0] // See the assemble.py script in ci-scripts https://github.com/foundriesio/ci-scripts/blob/18b4fb154c37b6ad1bc6e7b7903a540b7a758f5d/assemble.py#L300
```

Comment:
I think we need to set the `custom.apps_artifact` target attribute instead inside the ci-scripts; actually I think we need to set the `custom.ostree_artifact` target attribute too.  Otherwise, we're not enough safe.

Yes, I know that it means changes to the ci-scripts and would require the customer to switch to a new version of ci-scripts before using offline updates. But I don't think it creates any onerous problems for us.

The other option would be to not rely onto the jobserv and instead download these artifacts from the GCP buckets instead.  Not sure how feasible this approach is, however, as we allow user access to the hub.fio but not to the ostree-proxy.

One more thing to consider is how long we keep these artifacts on the jobserv shares? Do we prune them periodically? If yes - the current solution might not work as expected in long run.

Reply:
jobserv artifacts are in a gcp bucket, the api just gives you a signed url to them. and we don't delete them until a factory is deleted.

Reply:
> I think we need to set the `custom.apps_artifact` target attribute instead inside the ci-scripts; actually I think we need to set the `custom.ostree_artifact` target attribute too. Otherwise, we're not enough safe.

It makes sense to do something like that once we agree that we "really" need this feature.

Reply:
> The other option would be to not rely onto the jobserv and instead download these artifacts from the GCP buckets instead. Not sure how feasible this approach is, however, as we allow user access to the hub.fio but not to the ostree-proxy.

An ostree repo downloading via the ostree service/proxy looks quite feasible, not sure about Apps downloading since it depends on `skopeo` and secrets/auth to get access to 3rd party registries.